4.0.3 | Update xml-crypto

mcab opened the pull request:
See https://github.com/Clever/saml2/pull/285/commits/e5abdf32cd8656c7bbb66744eeaa1893bed900f2:

> Assertion digest checks are done on
> non-canonicalized documents, which could lead
> to unintended behaviors [1].
> 
> xml-crypto has backported a fix for this [2].
>
> [1] https://workos.com/blog/samlstorm
> [2] https://github.com/node-saml/xml-crypto/releases/tag/v3.2.1

<!--
Ceremony.

Ticket: [SECNG-3337]

**Testing:** This is a minor version bump with a backported fix on a version not officially supported by the developer. To my best knowledge, this should still work.

****
-->

[SECNG-3337]: https://clever.atlassian.net/browse/SECNG-3337?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mcab committed:
chore: Bump xml-crypto from ^3.0.1 to ^3.2.1

Assertion digest checks are done on
non-canonicalized documents, which could lead
to unintended behaviors [1].

xml-crypto has backported a fix for this [2].

[1] https://workos.com/blog/samlstorm
[2] https://github.com/node-saml/xml-crypto/releases/tag/v3.2.1

mcab committed:
4.0.3

- e5abdf3: Update xml-crypto to ^3.2.1.

mcab committed:
test: Ensure that newer LTS version (20) is tested

mcab committed:
bug: Match xml-crypto reference validation

`uri` may be parsed as undefined, which xml-crypto coerces into an empty string.

Match this behavior for consistency.

[1] https://github.com/node-saml/xml-crypto/compare/v3.2.0...v3.2.1#diff-58844ae8b1185b1627e00b2245f3fdcc5a8710438289e4555a0970ca8ae89fa6R528

felix-weizman-deel commented:
@mcab Thanks for fixing this! Do you have an estimate on when it will be merged?

mcab committed:
test: Add case for valid SAML response, commented out digest

This test case seems representative of what's expected from the library. The other
three tests cases in `xml-crypto`'s backported fix handle:

- invalid `SignedInfo` node (via loadSignature())
- additional nested `SignedInfo` node (via checkSignature())
- multiple `SignedInfo` nodes (via loadSignature())

All of these cases are handled within either checking or loading the signature.
This is handled by `xml-crypto`, which should gate check_saml_signature()'s
behavior.
